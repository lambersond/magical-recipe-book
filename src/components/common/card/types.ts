import type { MenuProps } from '../menu/types'
import type { MouseEventHandler } from 'react'

export type CardProps = {
  children: React.ReactNode
  className?: string
  onClick?: MouseEventHandler<HTMLButtonElement>
  overrideViews?: boolean
}

export type CardWithActionsProps = Omit<CardProps, 'children'> & {
  title: React.ReactNode
  children?: React.ReactNode
  options: MenuProps['options']
}

export type CollapsibleCardProps = {
  id?: string
  title: React.ReactNode
  subtitle?: React.ReactNode
  children: React.ReactNode
  classNames?: string
}
