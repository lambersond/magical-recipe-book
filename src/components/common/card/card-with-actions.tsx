import { Menu } from '../menu'
import { Popover } from '../popover'
import { Card } from './card'
import type { CardWithActionsProps } from './types'

export function CardWithActions({
  children,
  title,
  options,
  ...cardProps
}: CardWithActionsProps) {
  const showOptions = options.length > 0
  return (
    <Card {...cardProps}>
      <div className='flex justify-between items-start gap-2 w-full'>
        <div className='text-xl text-left font-bold break-all self-center'>
          {title}
        </div>
        {showOptions && (
          <Popover
            content={<Menu options={options} />}
            placement='top'
            asChild
            asKabab
          />
        )}
      </div>
      {children}
    </Card>
  )
}
