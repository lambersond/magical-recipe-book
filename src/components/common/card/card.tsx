import classNames from 'classnames'
import { card } from '@/utils/styles'
import type { CardProps } from './types'

export function Card({ children, className, onClick }: Readonly<CardProps>) {
  const classes = classNames(
    card,
    {
      'hover:border-hover cursor-pointer': !!onClick,
    },
    className,
  )

  if (onClick) {
    return (
      <button className={classes} onClick={onClick} data-testid='Card'>
        {children}
      </button>
    )
  }
  return (
    <div className={classes} data-testid='Card'>
      {children}
    </div>
  )
}
