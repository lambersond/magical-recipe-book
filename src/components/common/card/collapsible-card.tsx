'use client'

import { ChevronDown } from 'lucide-react'
import {
  expandablePaneClasses,
  updateStateAttributeAdvanced,
} from '@/utils/expandable'
import type { CollapsibleCardProps } from './types'

export function CollapsibleCard({
  id = 'collapsible-card',
  title,
  subtitle,
  children,
  classNames,
}: CollapsibleCardProps) {
  return (
    <section
      className={classNames}
      data-state='false'
      onClick={updateStateAttributeAdvanced(id)}
    >
      <div className='flex items-start justify-between cursor-pointer text-text-secondary'>
        <span className='flex flex-col flex-grow self-center'>
          {title}
          {subtitle}
        </span>
        <ChevronDown
          data-expandable
          className='data-[state=true]:rotate-180 transform transition-transform flex-shrink p-2 hover:text-text-primary hover:bg-hover rounded-full size-10 ml-2'
        />
      </div>
      <section
        data-expandable
        id={id}
        className={expandablePaneClasses}
        data-state='false'
        onClick={e => e.stopPropagation()}
      >
        {children}
      </section>
    </section>
  )
}
