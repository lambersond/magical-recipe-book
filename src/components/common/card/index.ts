export { Card } from './card'
export { CardWithActions } from './card-with-actions'
