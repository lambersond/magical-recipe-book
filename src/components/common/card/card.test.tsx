import { render, screen, useClick } from '@test-utils'
import { Card } from './card'

describe('components/common/card', () => {
  it('should match snapshot for clickable card', () => {
    expect(
      render(<Card onClick={jest.fn}>Hello World</Card>).asFragment(),
    ).toMatchSnapshot()
  })

  it('should match snapshot for basic card', () => {
    expect(render(<Card>Hello World</Card>).asFragment()).toMatchSnapshot()
  })

  it('should call onClick when clicked', async () => {
    const onClick = jest.fn()
    const click = useClick(<Card onClick={onClick}>Hello World</Card>)

    await click(screen.getByText('Hello World'))

    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
