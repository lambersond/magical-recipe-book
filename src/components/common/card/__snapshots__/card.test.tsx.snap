// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/common/card should match snapshot for basic card 1`] = `
<DocumentFragment>
  <div
    class="p-4 border-border bg-card w-full rounded-none sm:rounded-lg"
    data-testid="Card"
  >
    Hello World
  </div>
</DocumentFragment>
`;

exports[`components/common/card should match snapshot for clickable card 1`] = `
<DocumentFragment>
  <button
    class="p-4 border-border bg-card w-full rounded-none sm:rounded-lg hover:border-hover cursor-pointer"
    data-testid="Card"
  >
    Hello World
  </button>
</DocumentFragment>
`;
